src/core/mavsdk_impl.cpp

```diff
@@ -110,8 +110,71 @@ std::vector<std::shared_ptr<System>> MavsdkImpl::systems() const
     return systems_result;
 }
 
-void MavsdkImpl::receive_message(mavlink_message_t& message)
+void MavsdkImpl::forward_message(mavlink_message_t& message, Connection* connection)
+{
+    /**
+     * @brief forward_message Function implementing Mavlink routing rules.
+     * See https://mavlink.io/en/guide/routing.html
+     *
+     * This function was adapted from PX4 Firmware v1.11.3
+     * https://github.com/PX4/PX4-Autopilot/blob/v1.11.3/src/modules/mavlink/mavlink_main.cpp
+     */
+    const mavlink_msg_entry_t* meta = mavlink_get_msg_entry(message.msgid);
+
+    bool forward_heartbeats_enabled = false;
+    int target_system_id = 0;
+    int target_component_id = 0;
+
+    // might be nullptr if message is unknown
+    if (meta) {
+        // Extract target system and target component if set
+        if (meta->flags & MAV_MSG_ENTRY_FLAG_HAVE_TARGET_SYSTEM) {
+            target_system_id = (_MAV_PAYLOAD(&message))[meta->target_system_ofs];
+        }
+
+        if (meta->flags & MAV_MSG_ENTRY_FLAG_HAVE_TARGET_COMPONENT) {
+            target_component_id = (_MAV_PAYLOAD(&message))[meta->target_component_ofs];
+        }
+    }
+
+    // If it's a message only for us, we keep it, otherwise, we forward it.
+    const bool targeted_only_at_us =
+        (target_system_id == get_own_system_id() && target_component_id == get_own_component_id());
+
+    // We don't forward heartbeats unless it's specifically enabled.
+    const bool heartbeat_check_ok =
+        (message.msgid != MAVLINK_MSG_ID_HEARTBEAT || forward_heartbeats_enabled);
```

Comment:
I'm just wondering here. Should we forward heartbeats or not? The intuitive case would be to forward everything including heartbeats, right?

Reply:
Yes you are right, I think I wanted to leave it as an option, but then I forgot.
But maybe we should just enable it.

Reply:
I would enable it, and remove the specific code. And if we find out why it was there we can add it in :smile:.